File: jar-common/src/main/java/com/beancounter/common/model/IsoCurrencyPair.java
package com.beancounter.common.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.RequiredArgsConstructor;

@Data
@Builder
@EqualsAndHashCode
@RequiredArgsConstructor
@AllArgsConstructor
public class IsoCurrencyPair {
  private String from;
  private String to;

  @JsonIgnore
  public static IsoCurrencyPair from(Currency reporting, Currency trade) {
    if (trade == null || reporting == null) {
      return null;

    }
    if (trade.getCode().equalsIgnoreCase(reporting.getCode())) {
      return null;
    }

    return IsoCurrencyPair.builder()
        .from(reporting.getCode())
        .to(trade.getCode())
        .build();
  }

  @Override
  public String toString() {
    return from + ":" + to;
  }
}

File: jar-common/src/main/java/com/beancounter/common/model/IsoCurrencyPair.kt
package com.beancounter.common.model

data class IsoCurrencyPair(val from: String, val to: String) {

    override fun toString(): String {
        return "$from:$to"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as IsoCurrencyPair

        if (from != other.from) return false
        if (to != other.to) return false

        return true
    }

    override fun hashCode(): Int {
        var result = from.hashCode()
        result = 31 * result + to.hashCode()
        return result
    }

    companion object {
        @JvmStatic
        fun toPair(from: Currency, to: Currency): IsoCurrencyPair? {
            if (from.code == to.code) {
                return null
            }
            return IsoCurrencyPair(from.code, to.code)
        }
    }


}